File: Tutorials/Codes/Robot_Control_Course/Chapter16/Lesson3/Chapter16_Lesson3.py

import numpy as np

# Physical and control parameters
J = 0.05       # inertia
b = 0.01       # viscous damping
k_p = 50.0
k_d = 2.0

Ts = 0.001     # sampling period (s)
T_final = 5.0
N = int(T_final / Ts)

# Discrete-time matrices (forward Euler approximation)
A_d = np.array([
    [1.0, Ts],
    [-Ts * k_p / J, 1.0 - Ts * (k_d + b) / J]
])
B_d = np.zeros((2, 1))  # no explicit input when expressed in error coordinates

# Reference trajectory (constant)
q_ref = 1.0  # rad
e = np.array([0.0, 0.0])  # [position_error, velocity_error]

# Logging buffers
log_t = np.zeros(N)
log_e = np.zeros((N, 2))
log_u = np.zeros(N)

# Torque saturation
u_min, u_max = -5.0, 5.0

for k in range(N):
    t_k = k * Ts

    # Here q_meas, dq_meas would come from sensors (or a simulator)
    # For this simplified example, we reconstruct from error state:
    q_meas = q_ref - e[0]
    dq_meas = -e[1]

    # PD control law in joint coordinates
    u = -k_p * (q_meas - q_ref) - k_d * dq_meas

    # Saturation and logging
    u = float(np.clip(u, u_min, u_max))

    log_t[k] = t_k
    log_e[k, :] = e
    log_u[k] = u

    # Update error state using the derived matrix A_d
    e = A_d @ e

# Post-run: approximate discrete-time eigenvalues and check stability
eigvals = np.linalg.eigvals(A_d)
spectral_radius = max(abs(eigvals))

print("A_d:")
print(A_d)
print("Eigenvalues:", eigvals)
print("Spectral radius:", spectral_radius)

# Example invariant: energy-like quantity V_k = e^T P e
P = np.eye(2)
V = np.einsum("bi,ij,bj->b", log_e, P, log_e)  # V_k sequence

File: Tutorials/Codes/Robot_Control_Course/Chapter16/Lesson3/Chapter16_Lesson3.cpp

#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>
#include <Eigen/Dense>

using Eigen::Matrix2d;
using Eigen::Vector2d;

struct State {
  double q;
  double dq;
};

State readState();           // to be implemented for robot or simulator
void sendTorque(double u);   // to be implemented

int main() {
  const double J = 0.05;
  const double b = 0.01;
  const double k_p = 50.0;
  const double k_d = 2.0;

  const double Ts = 0.001;
  const double T_final = 5.0;
  const int N = static_cast<int>(T_final / Ts);

  // Discrete-time error dynamics matrix A_d (forward Euler)
  Matrix2d A_d;
  A_d << 1.0, Ts,
          -Ts * k_p / J, 1.0 - Ts * (k_d + b) / J;

  Vector2d e;  // [position_error, velocity_error]
  e.setZero();
  const double q_ref = 1.0;

  std::ofstream log("log_pd_debug.txt");

  for (int k = 0; k < N; ++k) {
    auto t_start = std::chrono::steady_clock::now();
    double t_k = k * Ts;

    State s = readState();  // q, dq from sensors or simulator
    e(0) = s.q - q_ref;
    e(1) = s.dq;

    double u = -k_p * e(0) - k_d * e(1);

    // Simple saturation
    const double u_min = -5.0;
    const double u_max =  5.0;
    if (u < u_min) u = u_min;
    if (u > u_max) u = u_max;

    sendTorque(u);

    // Logging: time, q, dq, e0, e1, u
    if (log.is_open()) {
      log << t_k << " "
          << s.q << " " << s.dq << " "
          << e(0) << " " << e(1) << " "
          << u << "\n";
    }

    // Next expected sampling time
    auto t_end = t_start + std::chrono::duration<double>(Ts);
    std::this_thread::sleep_until(t_end);
  }

  log.close();

  // Offline: eigenvalue check
  Eigen::EigenSolver<Matrix2d> es(A_d);
  std::cout << "A_d:\n" << A_d << std::endl;
  std::cout << "Eigenvalues:\n" << es.eigenvalues() << std::endl;

  return 0;
}
